spark/firmware
# Restores public server key and server address if missing

### Problem

There have been a number of community reports of devices losing their Public Server Key and/or Server Address.  Typically this is due to a power loss or reset during a critical flash write operation.

### Solution

- Public Server Key and Server Address are baked into System Firmware with this PR, and will be restored in RAM before a connection is attempted if either is found to be missing.
- This data will not be automatically written to FLASH when found missing to prevent further corruption.
- A cloud event will generated by the device alerting the cloud to which error is currently being patched by system firmware. (currently this does not hit the user's event stream)
- A subscription to a cloud event has been added to system firmware that will allow the user to restore the data to FLASH remotely when deemed appropriate. (currently there is no mechanism for the user to do this)
- The TCP server address in PSK is NULL terminated (backup copy in system firmware matches)
- The UDP server address in PSK is NOT NULL terminated (backup copy in system firmware will now include a NULL terminated server address).
- pubkey and private_key arrays in system_cloud_internal.cpp were not initialized and would contain garbage wherever key data was not written to, they are not initiallized to 0xFF to mimick erased flash memory.
- PR #1295 offers hardening and error checking to the FLASH read/write routines to help prevent this PR from being necessary, but just in case we will still be able to recover.

### Steps to Test

- All features have been manually tested on Photon/Electron/Core.
- Use dfu-util to erase [D] and verify [U] data at the appropriate address
	- Photon - dfu-util -d 2b04:d006 -a 1 -s 2082:768 -[D|U] blank.der
	- Electron - dfu-util -d 2b04:d00a -a 1 -s 3298:320 -[D|U] blank.der
	- Core - dfu-util -d 1d50:607f -a 1 -s 4096:768 -[D|U] blank.der
- Change system subscription/publish event name from `spark` to `spork` and publish PUBLIC event `spork/device/key/restore` to restore PSK and Server Address.

### Example App

This is an internal system change. No example application.

### References

- This will likely have a conflict with PR #1288 due to logging code where the public server address is read (prefer #1288's logging changes)
- Related to PR #1295

---

### Completeness


### Enhancement

- [[PR #1300]](https://github.com/spark/firmware/pull/1300) Restores public server key and server address if missing
